addons/api/addon/handlers/indexed-db-handler.js

```diff
@@ -0,0 +1,57 @@
+import { modelIndexes } from 'api/services/indexed-db';
+
+/**
+ * Handler to sit in front of the API request layer
+ * so we can handle any caching with indexedDb first
+ */
+const IndexedDbHandler = {
+  async request(context, next) {
+    switch (context.request.op) {
+      case 'query': {
+        const { store, data } = context.request;
+        const { type, query } = data;
+        const supportedModels = Object.keys(modelIndexes);
+
+        // TODO: Differentiate between user initiated paginated
+        //  requests and cache population requests
+
+        // Go through normal flow if we don't yet support the model
+        if (!supportedModels.includes(type)) {
+          break;
+        }
+
+        const adapter = store.adapterFor(type);
+        const schema = store.modelFor(type);
+
+        const payload = await adapter.query(store, schema, query);
+        const serializer = store.serializerFor(type);
+        const normalizedPayload = serializer.normalizeResponse(
+          store,
+          schema,
+          payload,
+          null,
+          'query'
+        );
+
+        const { data: payloadData } = normalizedPayload;
+
+        // TODO: Just to show how to add data to indexedDb
+        await this.indexedDb.db[type].bulkPut(payloadData);
```

Comment:
I've made the choice to actually store the normalized data instead of the raw JSON in indexed DB, any concerns with this?

Reply:
Was there something that caused you to change your approach? Are there any cons to storing the normalized data?

Reply:
The raw JSON isn't an accurate representation of what data the user interacts with, we'd still need to handle serialization/normalization each time we query for data which would complicate searching and filtering. 

I also want to be able to use the data as is and move it into the ember data store.

Reply:
That's right. You just didn't want the hydrated models. I am down with normalizing it first.